network/lwip2transport/udp.go

```diff
@@ -15,103 +15,95 @@
-	// is closed.
-	timeout time.Duration
-
-	// Maps local UDP addresses (IPv4:port/[IPv6]:port) to connections to the proxy.
-	conns map[string]net.PacketConn
+	mu       sync.Mutex                       // Protects the sessions field
+	handler  network.PacketHandler            // A network stack neutral implementation of UDP handler
+	sessions map[string]network.PacketSession // Maps local lwIP UDP socket (IPv4:port/[IPv6]:port) to PacketSession
+	timeout  time.Duration                    // Session timeout is how long a session can last before closing
 }
 
 // newUDPHandler returns a lwIP UDP connection handler.
 //
-// `pl` is an UDP packet listener.
+// `h` is a handler that handles UDP packets.
 // `timeout` is the UDP read and write timeout.
-func newUDPHandler(pl transport.PacketListener, timeout time.Duration) *udpHandler {
+func newUDPHandler(h network.PacketHandler, timeout time.Duration) *udpHandler {
 	return &udpHandler{
-		listener: pl,
+		handler:  h,
 		timeout:  timeout,
-		conns:    make(map[string]net.PacketConn, 8),
+		sessions: make(map[string]network.PacketSession, 8),
 	}
 }
 
-func (h *udpHandler) Connect(tunConn lwip.UDPConn, target *net.UDPAddr) error {
-	proxyConn, err := h.listener.ListenPacket(context.Background())
+// Connect creats a new UDP session from the [transport.PacketServer]. It also schedule a timer with the timeout of the
+// session; the session will be closed once the timer triggers.
+func (h *udpHandler) Connect(tunConn lwip.UDPConn, _ *net.UDPAddr) error {
+	laddr := tunConn.LocalAddr().String()
+
+	// Lock the entire function to prevent from creating multiple sessions of one laddr
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	if _, ok := h.sessions[laddr]; ok {
+		return fmt.Errorf("duplicated connection %v", laddr)
+	}
+
+	// Try to create a new session
+	respw := &udpConnResponseWriter{tunConn}
+	session, err := h.handler.NewSession(tunConn.LocalAddr(), respw)
 	if err != nil {
+		respw.Close()
 		return err
 	}
-	h.Lock()
-	h.conns[tunConn.LocalAddr().String()] = proxyConn
-	h.Unlock()
-	go h.relayPacketsFromProxy(tunConn, proxyConn)
-	return nil
-}
+	h.sessions[laddr] = session
 
-// relayPacketsFromProxy relays packets from the proxy to the TUN device.
-func (h *udpHandler) relayPacketsFromProxy(tunConn lwip.UDPConn, proxyConn net.PacketConn) {
-	buf := lwip.NewBytes(lwip.BufSize)
-	defer func() {
-		h.close(tunConn)
-		lwip.FreeBytes(buf)
-	}()
-	for {
-		proxyConn.SetDeadline(time.Now().Add(h.timeout))
-		n, sourceAddr, err := proxyConn.ReadFrom(buf)
-		if err != nil {
-			return
-		}
-		// No resolution will take place, the address sent by the proxy is a resolved IP.
-		sourceUDPAddr, err := net.ResolveUDPAddr("udp", sourceAddr.String())
-		if err != nil {
-			return
-		}
-		_, err = tunConn.WriteFrom(buf[:n], sourceUDPAddr)
-		if err != nil {
-			return
-		}
-	}
+	// Clean-up the session after timeout
+	time.AfterFunc(h.timeout, func() {
```

Comment:
This is no longer an idle timeout. You are setting a fixed time out and never updating it. It will remove sessions on the fixed timeout regardless of use. That's not what we want.

The original logic was updating the timeout on every packet. It was measuring idle timeout. We need to restore that behavior.

I believe a good approach would be to put that logic in a handler though. On request (or response), you update the timeout. When the timeout expires, we close the ResponseWriter. That should tell the lwip handler to remove the association.

With that, people can provide their own session expiration logic. For example, in the server we expire DNS requests faster than other connections, since we have many of them. We could possibly reuse the same logic here.

Reply:
Implemented timeout in `packetListenerProxyAdapter`